direktiv/direktiv-ui
# [DIR-194] Monitoring Page

# This PR introduces the monitoring page in the new design

**[Preview 🚀](https://direktiv-ui-refs-pull-440-merge.direktiv.dev/)**

![CleanShot 2023-08-10 at 09 38 41@2x](https://github.com/direktiv/direktiv-ui/assets/121789579/6911621f-2202-4e57-858c-63705963954b)

## Other Changes

Playwright tests have been re-enabled on every Pull Request once again.

I replaced the [Bug](https://lucide.dev/icons/bug) icon with the [ActivitySquare](https://lucide.dev/icons/activity-square)

Used the `TooltipCopyBadge` on the invoker column on the instance page as well (only for child instance)

fixes a bug in the Button component where the following code will not behave like a link (button is not clickable because the class `pointer-events-none` was accidentally applied)

```typescript
<Button asChild>
  <Link to="...">Link</Link>
</Button>
``` 
<details><summary>On the instances Page, the full screen button on the diagram panel had no background</summary>

| Before | After |
|--------|--------|
| ![CleanShot 2023-08-09 at 15 52 39@2x](https://github.com/direktiv/direktiv-ui/assets/121789579/8821d150-dc21-4e37-8653-f600dc584a89) | ![CleanShot 2023-08-09 at 15 52 47@2x](https://github.com/direktiv/direktiv-ui/assets/121789579/c4634d0f-ac7a-49d5-9604-35df9276a869) | 
</details> 

<details><summary>Added a new RefreshButton component for better feedback when the user clicked a refresh button</summary>

![CleanShot 2023-08-10 at 08 42 37](https://github.com/direktiv/direktiv-ui/assets/121789579/2d3f589f-2e70-4f89-9926-1d6c68dbf7eb)

</details> 

<details><summary>I changed the pattern on how we consume the streaming data in our components</summary>
<p>

The old API looked like this:

```typescript
...
const InstancesPage = () => {
  const { t } = useTranslation();

  /**
   * This hook will subscribe to a SSE stream. It will setup a connection
   * on mount and cancel the connection on unmount. To avoid unnecessary
   * reconnects, make sure to place this hooks in a isolated a parent component
   * that will not rerender very often.
   */
  useNamespaceLogsStream();

  return (
    <div className="flex grow flex-col gap-y-4 p-5">
      <h3 className="flex items-center gap-x-2 font-bold">
        <ActivitySquare className="h-5" />
...
``` 

We offer a hook for streaming, but it comes with a separate handbook on how to use it. When you place that hook wrong, it will reconnect a lot. Using streaming should be more simple and the context should not matter at all. The new solution lets us place a streaming provider component where ever we want (we can even place it deep down in a tree, when one of the parents is affected by the streaming updates and needs to re-render)

The problem of the previous solution was, that it was dependent on where it was placed. Since it was a hook it was re-rendering a lot, when the component it was in has changed. My idea was to somehow memo-ize the hook to avoid unnecessary re-renderings. But the hook has to have a `useEffect` in it and I did not find a good solution to memo-ize an effect (I think it is simply not possible without violating the exhaustive-deps eslint rule). 

So the solution would be to place it in its own component and put a `memo()` around it to only re-render when the props change. The new usage looks like this.

```typescript
...
const InstancesPage = () => {
  const { t } = useTranslation();
  return (
    <div className="flex grow flex-col gap-y-4 p-5">
      <NamespaceLogsStreamingProvider />  {/* could also be  <NamespaceLogsStreamingProvider enabled={somevar} /> */}
       ...
    </div>
...
```
My first intuition was to have it like this `<NamespaceLogsStreamingProvider>...</NamespaceLogsStreamingProvider>` but this would also re-render too often, because children are just props as well and memo will re-render when the children change. And it would also be kind of wrong to make this a wrapping component, because it would look like it would do something for its children while it does not have any parent child dependency. An empty component like `<NamespaceLogsStreamingProvider />` looks a bit weird and looks like it should be a hook, but I did not find a way to replicate the same behavior in a hook. 

Here is a screen recording of my test case:

https://github.com/direktiv/direktiv-ui/assets/121789579/dde1fd01-e4de-4771-96f1-3012d91e0917

- I put the provider directly inside the log component (which re-renders a lot when new logs comes in)
- I made a toggle for the `enabled` attribute of `NamespaceLogsStreamingProvider` to demo that the component will re-render when it has to (when props change)
- the network tab and the console output in this video shows when a new stream connection is established and when it is closed
- The old method would have shown a new connection in the network tab whenever we click that toggle, and also whenever a new log comes in. 
- the new method is able to avoid re-rendering when anything around it does, but it will rerender when it has to (e.g. when its own prop changes)
- I wonder if we should call it "provider". My mental model of a provider component would provide something to its children. What do you think about the name?
</p>
</details> 

## How to test

A good way to test the logs is to have two windows open (use two different browsers just to make sure you don't hit the browser limit for maximal number of open SSE connections) and watch the monitoring page as you trigger a workflow that spawns a lot of children automatically

## follow-up tickets
- [E2E test](https://linear.app/direktiv/issue/DIR-735/test-monitoring-page)
